Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/safetynet/CWASafetyNet.kt

```diff
@@ -1,19 +1,93 @@
+    private val googleApiVersion: GoogleApiVersion,
+    private val cwaSettings: CWASettings,
+    private val timeStamper: TimeStamper
+) : DeviceAttestation {
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    fun generateSalt(): ByteArray = ByteArray(16).apply {
+        secureRandom.nextBytes(this)
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    fun calculateNonce(salt: ByteArray, payload: ByteArray): String {
+        val concat = salt + payload
+        return concat.toSHA256()
+    }
+
     override suspend fun attest(request: DeviceAttestation.Request): DeviceAttestation.Result {
-        return object : DeviceAttestation.Result {
-            override val accessControlProtoBuf: PpacAndroid.PPACAndroid = PpacAndroid.PPACAndroid.getDefaultInstance()
+        if (!googleApiVersion.isPlayServicesVersionAvailable(13000000)) {
+            throw SafetyNetException(Type.PLAY_SERVICES_VERSION_MISMATCH, "Google Play Services too old.")
+        }
 
-            override fun requirePass(requirements: SafetyNetRequirements) {
-                // Passed
+        appConfigProvider.getAppConfig().apply {
+            if (deviceTimeState == ConfigData.DeviceTimeState.ASSUMED_CORRECT) {
+                throw SafetyNetException(Type.DEVICE_TIME_UNVERIFIED, "Device time is unverified")
+            }
+            if (deviceTimeState == ConfigData.DeviceTimeState.INCORRECT) {
+                throw SafetyNetException(Type.DEVICE_TIME_INCORRECT, "Device time is incorrect")
             }
         }
+
+        val firstReliableTimeStamp = cwaSettings.firstReliableDeviceTime
+        if (firstReliableTimeStamp == Instant.EPOCH) {
+            throw SafetyNetException(Type.TIME_SINCE_ONBOARDING_UNVERIFIED, "No first reliable timestamp available")
+        } else if (Duration(firstReliableTimeStamp, timeStamper.nowUTC) < Duration.standardHours(24)) {
+            throw SafetyNetException(Type.TIME_SINCE_ONBOARDING_UNVERIFIED, "Time since first reliable timestamp <24h")
+        }
+
+        val salt = generateSalt()
+        val nonce = calculateNonce(salt = salt, payload = request.scenarioPayload)
+        Timber.tag(TAG).d("With salt=%s and payload=%s, we created none=%s", salt, request.scenarioPayload, nonce)
+
+        val report = client.attest(nonce.toByteArray())
+
+        report.error?.let {
+            Timber.tag(TAG).w("SafetyNet Response has an error message: %s", it)
+        }
+
+        if (nonce != report.nonce) {
+            throw SafetyNetException(
+                Type.NONCE_MISMATCH,
+                "Request nonce doesn't match response ($nonce != ${report.nonce})"
+            )
+        }
+
+        if (context.packageName != report.apkPackageName) {
+            throw SafetyNetException(
+                Type.APK_PACKAGE_NAME_MISMATCH,
+                "Our APK name doesn't match response (${context.packageName} != ${report.apkPackageName})"
+            )
+        }
+
```

Comment:
Quote from this [article](https://android-developers.googleblog.com/2017/11/10-things-you-might-be-doing-wrong-when.html)

> ... and use timestampMs to check how much time has passed since you made the request and you got the response. A delayed response that arrives several hours or days after the request may suggest suspicious activity.

Should we check the timestamp too?

Reply:
Okay, the TechSpec says that this is done on the server.